src/mlpack/tests/main_tests/linear_svm_test.cpp

```diff
@@ -0,0 +1,893 @@
+{
+  arma::mat trainData1({{1.0, 2.0, 3.0}, {1.0, 4.0, 9.0}, {0, 1, 1}});
+  arma::mat testData({{4.0, 5.0}, {1.0, 6.0}});
+
+  SetInputParam("training", std::move(trainData1));
+  SetInputParam("test", testData);
+
+  // The first solution.
+  mlpackMain();
+
+  // Get the output.
+  const arma::Row<size_t> testLabels1 =
+      std::move(CLI::GetParam<arma::Row<size_t>>("predictions"));
+
+  // Reset the settings.
+  bindings::tests::CleanMemory();
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+
+  // Now train by providing labels as extra parameter.
+  arma::mat trainData2({{1.0, 2.0, 3.0}, {1.0, 4.0, 9.0}});
+  arma::Row<size_t> trainLabels({0, 1, 1});
+
+  SetInputParam("training", std::move(trainData2));
+  SetInputParam("labels", std::move(trainLabels));
+  SetInputParam("test", std::move(testData));
+
+  // The second solution.
+  mlpackMain();
+
+  // get the output
+  const arma::Row<size_t> &testLabels2 =
+      CLI::GetParam<arma::Row<size_t>>("predictions");
+
+  // Both solutions should be equal.
+  CheckMatrices(testLabels1, testLabels2);
+}
+
+/**
+ * Ensure that saved model can be used again.
+ */
+BOOST_AUTO_TEST_CASE(LinearSVMModelReuseTest)
+{
+  arma::mat trainData;
+  if (!data::Load("iris.csv", trainData))
+    BOOST_FAIL("Cannot load test dataset iris.csv!");
+
+  arma::Row<size_t> trainLabels;
+  if (!data::Load("iris_labels.txt", trainLabels))
+    BOOST_FAIL("Cannot load test dataset iris_labels.txt!");
+
+  arma::mat testData;
+  if (!data::Load("iris_test.csv", testData))
+    BOOST_FAIL("Cannot load test dataset iris_test.csv!");
+
+  SetInputParam("training", std::move(trainData));
+  SetInputParam("labels", std::move(trainLabels));
+  SetInputParam("test", testData);
+
+  // First solution
+  mlpackMain();
+
+  // Get the output model obtained from training.
+  LinearSVM<>* model =
+      CLI::GetParam<LinearSVM<>*>("output_model");
+  // Get the output.
+  const arma::Row<size_t> &testLabels1 =
+      std::move(CLI::GetParam<arma::Row<size_t>>("predictions"));
+
+  // Reset the data passed.
+  CLI::GetSingleton().Parameters()["training"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["labels"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["test"].wasPassed = false;
+
+  SetInputParam("input_model", model);
+  SetInputParam("test", std::move(testData));
+
+  // Second solution.
+  mlpackMain();
+
+  // Get the output.
+  const arma::Row<size_t> &testLabels2 =
+      CLI::GetParam<arma::Row<size_t>>("predictions");
+
+  // Both solutions should be equal.
+  CheckMatrices(testLabels1, testLabels2);
+}
+
+/**
+  * Checking for dimensionality of the test data set.
+ **/
+BOOST_AUTO_TEST_CASE(LinearSVMCheckDimOfTestData)
+{
+  arma::mat trainData;
+  if (!data::Load("iris.csv", trainData))
+    BOOST_FAIL("Cannot load test dataset iris.csv!");
+
+  arma::mat testData;
+  if (!data::Load("iris_test.csv", testData))
+    BOOST_FAIL("Cannot load test dataset iris_test.csv!");
```

Comment:
I think the dimensionalities are actually the same here for these files, but the key is that you're not specifying labels, so `mlpackMain()` assumes the last dimension is the labels, thus causing the model to be trained with `trainData.n_rows - 1` rows.  That isn't obvious from just looking at the code here---can you add a comment to point this out?

Reply:
Add a comment for this.

Reply:
Thanks. :+1: